core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -73,6 +104,33 @@ public void request(int numMessages) {
   public void sendHeaders(Metadata headers) {
     checkState(!sendHeadersCalled, "sendHeaders has already been called");
     checkState(!closeCalled, "call is closed");
+
+    headers.removeAll(MESSAGE_ENCODING_KEY);
+    if (compressor == null) {
+      compressor = Codec.Identity.NONE;
+      if (inboundHeaders.containsKey(MESSAGE_ACCEPT_ENCODING_KEY)) {
+        String acceptEncodings = inboundHeaders.get(MESSAGE_ACCEPT_ENCODING_KEY);
+        for (String acceptEncoding : ACCEPT_ENCODING_SPLITER.split(acceptEncodings)) {
+          Compressor c = compressorRegistry.lookupCompressor(acceptEncoding);
+          if (c != null) {
+            compressor = c;
+            break;
+          }
+        }
+      }
+    }
+    inboundHeaders = null;
+    if (compressor != Codec.Identity.NONE) {
+      headers.put(MESSAGE_ENCODING_KEY, compressor.getMessageEncoding());
```

Comment:
Need to verify that the client supports the compression type chosen. (the comment was collapsed)